### PR TITLE
RF: weight parametrization

### LINKED ISSUE #1518: RF weight dropout and variational noise
Currently we don't have weight dropout in the RF. We should add it.

(I thought there was an issue already about it but I don't find it.)

Related:

* Weight dropout in RETURNN common: `returnn_common.nn.utils.weight_dropout.weight_dropout`. Related issue: https://github.com/rwth-i6/returnn_common/issues/59, https://github.com/rwth-i6/returnn_common/issues/250
* Weight dropout for TF net-dict. See: https://github.com/rwth-i6/returnn/issues/735
* Related discussion (on weight norm): https://github.com/rwth-i6/returnn/issues/1264#issuecomment-1446276707

In general, there is a whole class of similar features on parameter reparameterizations, which would require a similar mechanism, like weight norm.

Regarding implementation:

I think we could follow the PyTorch implementation of similar logic (e.g. weight norm) by using a forward-pre-hook. We already have support for hooks via `rf.Module.register_forward_hook`/`rf.hooks.setup_post_hook_on_method`.

Regarding PyTorch:

* [`torch.nn.utils.weight_norm`](https://pytorch.org/docs/stable/generated/torch.nn.utils.weight_norm.html) uses `register_forward_pre_hook`. However, the doc says:
  > This function is deprecated. Use [torch.nn.utils.parametrizations.weight_norm()](https://pytorch.org/docs/stable/generated/torch.nn.utils.parametrizations.weight_norm.html#torch.nn.utils.parametrizations.weight_norm) which uses the modern parametrization API.
* [`torch.nn.utils.parametrizations.weight_norm`](https://pytorch.org/docs/stable/generated/torch.nn.utils.parametrizations.weight_norm.html#torch.nn.utils.parametrizations.weight_norm) uses the modern parametrization API, i.e. [`torch.nn.utils.parametrize.register_parametrization`](https://pytorch.org/docs/stable/generated/torch.nn.utils.parametrize.register_parametrization.html#torch.nn.utils.parametrize.register_parametrization). On `register_parametrization` (it's somewhat ugly (although our own hooks mechanism is also a bit complicated), but this here looks even worse):
  * It uses `_inject_new_class` to replace the original module class by a new dummy one, which is going to be extended by adding a property for the parameter.
    (We cannot add the property to the object. A property can only be added to a class. See [the descriptor guide doc](https://docs.python.org/3/howto/descriptor.html).)
  * It deletes the parameter (`delattr`) and stores it in a `ParametrizationList`.
  * It injects a property (`_inject_property`) for the parameter. This will call the `parametrization` to get it.
  * Note that this implementation of parametrization is not working with scripting. (I'm not exactly sure why though?)
  * Even tracing is not working together with caching? (Again not sure why?)
  * There is a caching mechanism which is disabled by default. The user need to explicitly enable it temporarily, like so:
    ```python
    import torch.nn.utils.parametrize as P
    ...
    with P.cached():
        output = model(inputs)
    ```
    When leaving the `cached()` context, it will clear the (global) cache.

So, comparing the modern parametrization API `register_parametrization` to the way it was done in `torch.nn.utils.weight_norm` via `register_forward_pre_hook`:
* `register_parametrization` also works for any other module methods, while `register_forward_pre_hook` only works for `forward`.
* `register_forward_pre_hook` will create a temporary buffer to store the calculated parameter. This is buffer is not freed after `forward`. It is just updated in the next pre-hook. Via `register_parametrization`, the calculated parameter is freed after it's usage (and with caching enabled, after you leave the `cached()` context scope). So `register_parametrization` requires less memory.
* `register_parametrization` will recompute the parameter several times when it is accessed multiple times (whenever the property is accessed), unless the user uses the parameterization `cached` mechanism explicitly. But you usually would want to avoid the redundant computation, i.e. you want to use the `cached` context. But that means you also need to modify the way you call your model by installing the `cached` context somewhere. This is not automatic. On the other side, `register_forward_pre_hook` is always automatic. (Although, if you call `forward` multiple times, it would also cause redundant recomputations of the parameter.)
* `register_forward_pre_hook` is conceptually much simpler and less ugly than `register_parametrization`.
* PyTorch parametrization API introduction and discussions: PR https://github.com/pytorch/pytorch/pull/33344, issue https://github.com/pytorch/pytorch/issues/28937, issue https://github.com/pytorch/pytorch/issues/7313.

Concluding from that, I'm a bit unsure what way to go for RF... Using `register_forward_pre_hook` looks too error-prone (only covers the hooked function, nothing else)... but the other approach looks too complicated? But maybe still better. The caching mechanism is maybe also not so important for now? For all use cases, I think it would not matter (e.g. `rf.Linear`, `rf.SelfAttention`, `rf.Conv`, etc.).

Further, we should also support this with gradient checkpointing such that the weights are not stored twice in memory. In our existing TF implementation of variational noise, we already use gradient checkpointing, where only the random number generator state is stored and not the dropout mask nor the weight. Thus there is almost no memory overhead. See `gradient_checkpoint_scope` and co. For PyTorch, it is currently unclear how to do this. I moved this over to a separate issue: #1552

### PR DESCRIPTION
**Pull request by albertz**
Fix #1518.

* Add `rf.gradient_checkpoint_scope`.
* Add `rf.register_parametrization`, similarly as in PyTorch.
* Add `rf.weight_dropout` and `rf.weight_noise`.

### CODE REVIEW

**Comment by albertz** [comment]
Still WIP. Well, the parametrization API is basically already ready. But I want to add some tests for it. Also I want to add weight dropout and variational noise already. Maybe also weight norm. Similar as here:
https://pytorch.org/docs/stable/_modules/torch/nn/utils/parametrizations.html
https://pytorch.org/docs/stable/generated/torch.nn.utils.parametrizations.weight_norm.html#torch.nn.utils.parametrizations.weight_norm

I'm probably not squashing all commits together but leave them as separate commits. This PR here is mostly just for easier reviewing, and maybe also discussions. (Although the main design questions are already being discussed in #1518.)

**Comment by albertz** [comment]
Ok, all done, at least `weight_dropout` and `weight_noise`. I think this is enough for this PR.

**Comment by albertz** [comment]
I already merged this because I want to use this in some experiments, and I think we can just directly fix any issues in the code in follow-up commits/PRs. I think we even can still do changes to the API for a short while, as long as no-one else is using this.

So, please anyway review/comment.
